### PR TITLE
feat(login): add component and  unit test

### PR DESCRIPTION
**Pull request by YgRamos9**
## Summary
- add component
- add unit test
- add validations

## Screenshot
Captura de Pantalla 2022-08-04 a la(s) 12 43 21 p m

## Trello Card
https://trello.com/c/cSyO6LU5/18-login

### CODE REVIEW

File: angular/src/app/components/form-container/form-container.component.html [hunk 1]
```diff
@@ -9,5 +9,9 @@
     </div>
     <input type="submit" [value]="buttonSend" class="button" />
   </form>
+  <hr class="line-form" />
   <ng-content></ng-content>
 </div>
+<div *ngIf="mensajeError" class="alert danger-alert">
```
Comment (joacomiguens) [inline angular/src/app/components/form-container/form-container.component.html]:
```suggestion
<div *ngIf="errorMessage" class="alert danger-alert">
```

File: angular/src/app/services/user.service.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ export class UserService {
 
   createUser(user: IUser): Observable<HttpResponse<object>> {
     user.locale = 'en';
+    const headers = 'headers made with HttpHeaders';
```
Comment (joacomiguens) [inline angular/src/app/services/user.service.ts]: why? 🤔

File: angular/src/styles/commons/alerts.scss [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  border-radius: 5px;
+  box-shadow: 0 0 15px 5px $silver;
+  animation-duration: 5s;
+  animation-name: slidein;
+  position: fixed;
+  top: 0;
+  right: 0;
+  transform: translateX(100%);
+  justify-content: center;
+}
```
Comment (joacomiguens) [inline angular/src/styles/commons/alerts.scss]: sort

File: angular/src/app/app.module.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
-import { HttpClientModule } from '@angular/common/http';
+import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
```
Comment (joacomiguens) [inline angular/src/app/app.module.ts]: why interceptors here?

File: angular/src/app/screens/login/login.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import { APP_BASE_HREF } from '@angular/common';
+import { fireEvent, render, waitFor } from '@testing-library/angular';
+import { screen } from '@testing-library/dom';
+import { LOGIN_FIELDS } from '../../constants/formAccount';
+import { LoginComponent } from './login.component';
+import { LoginRoutingModule } from './login-routing.module';
+import '@testing-library/jest-dom';
+
+describe('FormContainerComponent', () => {
+  const createUserMock = jest.fn();
```
Comment (joacomiguens) [inline angular/src/app/screens/login/login.component.spec.ts]: createUserMock or loginUserMock? 🤔

File: angular/src/app/screens/login/login.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      },
+      providers: [{ provide: APP_BASE_HREF, useValue: '/my/app' }],
+    });
+  });
+
+  it('check all empty fields', async () => {
+    const form = screen.getByRole('form');
+    fireEvent.submit(form);
+    const errorSpans = screen.getAllByTestId('error-span');
+    expect(errorSpans).toHaveLength(2);
```
Comment (joacomiguens) [inline angular/src/app/screens/login/login.component.spec.ts]: Check real messages to be 100% sure that you are showing what you want.

File: angular/src/app/screens/login/login.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  });
+
+  it('check fields with incorrect validation', async () => {
+    const email = screen.getByLabelText(/Email/i);
+    fireEvent.input(email, {
+      target: {
+        value: 'ramos111com12',
+      },
+    });
```
Comment (joacomiguens) [inline angular/src/app/screens/login/login.component.spec.ts]: use addValueEvent

File: angular/src/app/screens/login/login.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import { CommonModule } from '@angular/common';
```
Comment (joacomiguens) [inline angular/src/app/screens/login/login.component.spec.ts]: You should check what happened when there is a missing field and when credentials are wrong 💯

File: angular/src/app/services/user.service.ts (hunk 2 above)
Comment (YgRamos9) [reply angular/src/app/services/user.service.ts]: excuse me please, it was a mistake on my part I did not remove the constant after testing with the interceptor

File: angular/src/app/screens/login/login.component.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import { UserService } from '../../services/user.service';
+
+@Component({
+  selector: 'app-login',
+  templateUrl: './login.component.html',
+})
+export class LoginComponent {
+  fields = LOGIN_FIELDS;
+  buttonSend = 'Login';
+  mensajeError!: string;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.component.ts]:
```suggestion
  errorMessage!: string;
```

File: angular/src/app/screens/login/login.component.html [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<app-form-container [formFields]="fields" (handleOnSubmit)="handleOnSubmit($event)" [buttonSend]="buttonSend"
+  [mensajeError]="mensajeError">
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.component.html]: same

File: angular/src/app/components/form-container/form-container.component.ts [hunk 11]
```diff
@@ -8,26 +8,37 @@ import { IFields, IUser } from 'src/app/interfaces/global.interface';
   templateUrl: './form-container.component.html',
   styleUrls: ['./form-container.component.scss'],
 })
-export class FormContainerComponent {
+export class FormContainerComponent implements OnInit {
   @Input() formFields!: IFields[];
   @Input() buttonSend!: string;
   @Output() handleOnSubmit = new EventEmitter<IUser>();
+  @Input() mensajeError!: string;
```
Comment (miguelhenaoa) [inline angular/src/app/components/form-container/form-container.component.ts]: same

File: angular/src/app/constants/formAccount.ts [hunk 12]
```diff
@@ -22,7 +22,7 @@ export const SIGN_UP_FIELDS = [
     label: 'Password',
     type: 'password',
     errorMessage:
-      'El password es requerido y debe contener al menos 1 mayúscula, 1 minúscula, 1 número y 1 caracter',
+      'La password es requerido y debe contener al menos 1 mayúscula, 1 minúscula, 1 número y 1 caracter',
```
Comment (miguelhenaoa) [inline angular/src/app/constants/formAccount.ts]: If you use the pronoun **la**, change 'requerido' to 'requerida', or better change 'password' to 'contraseña'.
```suggestion
      'La contraseña es requerida y debe contener al menos 1 mayúscula, 1 minúscula, 1 número y 1 carácter',
```

File: angular/src/app/constants/formAccount.ts [hunk 13]
```diff
@@ -31,3 +31,19 @@ export const SIGN_UP_FIELDS = [
...
+    label: 'Email',
+    type: 'email',
+    errorMessage: 'El formato de mail no es correcto',
+  },
+  {
+    key: 'password',
+    label: 'Password',
+    type: 'password',
+    errorMessage:
+      'La password es requerido y debe contener al menos 1 mayúscula, 1 minúscula, 1 número y 1 caracter',
```
Comment (miguelhenaoa) [inline angular/src/app/constants/formAccount.ts]: Same

File: angular/src/app/screens/login/login.module.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+import { FormContainerModule } from 'src/app/components/form-container/form-container.module';
+@NgModule({
+  declarations: [LoginComponent],
+  imports: [
+    CommonModule,
+    LoginRoutingModule,
+    ReactiveFormsModule,
+    FormContainerModule,
+  ],
+  exports: [LoginComponent, CommonModule, ReactiveFormsModule],
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.module.ts]: The same, it is not necessary to export modules or components.

File: angular/src/app/constants/codeRequest.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+export enum Status {
+  success = 201,
```
Comment (miguelhenaoa) [inline angular/src/app/constants/codeRequest.ts]:
```suggestion
  Created = 201,
```

File: angular/src/app/constants/codeRequest.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+  ok = 200,
+  unAuthorized = 401,
```
Comment (miguelhenaoa) [inline angular/src/app/constants/codeRequest.ts]:
```suggestion
  Unauthorized = 401,
```

File: angular/src/app/constants/codeRequest.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+export enum Status {
+  success = 201,
+  ok = 200,
```
Comment (miguelhenaoa) [inline angular/src/app/constants/codeRequest.ts]:
```suggestion
  Ok= 200,
```

File: angular/src/app/constants/codeRequest.ts [hunk 18]
```diff
@@ -0,0 +1,5 @@
+export enum Status {
```
Comment (miguelhenaoa) [inline angular/src/app/constants/codeRequest.ts]: For enums use UpperCamelCase

File: angular/src/app/screens/login/login.component.ts [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+import { UserService } from '../../services/user.service';
+
+@Component({
+  selector: 'app-login',
+  templateUrl: './login.component.html',
+})
+export class LoginComponent {
+  fields = LOGIN_FIELDS;
+  buttonSend = 'Login';
+  errorMessage!: string;
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.component.ts:17]: blank line

**Review by miguelhenaoa — CHANGES_REQUESTED** [review]

File: angular/src/app/screens/login/login.component.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+  constructor(private userService: UserService) {}
+
+  formControls = new FormGroup({
+    email: new FormControl('', [Validators.required, Validators.email]),
+    password: new FormControl('', [
+      Validators.required,
+      Validators.pattern('(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9]).{6,}'),
```
Comment (joacomiguens) [inline angular/src/app/screens/login/login.component.ts]: Abstract regex to a constant, and use it on login and sign-up components

File: angular/src/app/screens/login/login.module.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { LoginComponent } from './login.component';
+import { ReactiveFormsModule } from '@angular/forms';
+import { LoginRoutingModule } from './login-routing.module';
+import { FormContainerModule } from 'src/app/components/form-container/form-container.module';
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.module.ts]: Sort

File: angular/src/app/screens/sign-up/sign-up.component.ts [hunk 22]
```diff
@@ -1,25 +1,41 @@
 import { HttpResponse } from '@angular/common/http';
 import { Component } from '@angular/core';
-import { SIGN_UP_FIELDS } from '../../constants/formAccount';
+import { SIGN_UP_FIELDS } from '../../constants/form-account';
 import { IUser } from 'src/app/interfaces/global.interface';
 import { UserService } from '../../services/user.service';
+import { StatusRequest } from 'src/app/constants/code-request';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { CustomValidators } from 'src/app/helpers/utilities/custom-validators';
```
Comment (miguelhenaoa) [inline angular/src/app/screens/sign-up/sign-up.component.ts]: Sort

File: angular/src/app/screens/sign-up/sign-up.component.ts [hunk 23]
```diff
@@ -1,25 +1,41 @@
 import { HttpResponse } from '@angular/common/http';
 import { Component } from '@angular/core';
-import { SIGN_UP_FIELDS } from '../../constants/formAccount';
+import { SIGN_UP_FIELDS } from '../../constants/form-account';
 import { IUser } from 'src/app/interfaces/global.interface';
 import { UserService } from '../../services/user.service';
+import { StatusRequest } from 'src/app/constants/code-request';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
```
Comment (joacomiguens) [inline angular/src/app/screens/sign-up/sign-up.component.ts]: must be at top

File: angular/src/app/screens/sign-up/sign-up.component.ts [hunk 24]
```diff
@@ -1,25 +1,41 @@
...
   constructor(private userService: UserService) {}
 
+  formControls = new FormGroup(
+    {
+      firstName: new FormControl('', [Validators.required]),
+      lastName: new FormControl('', [Validators.required]),
+      email: new FormControl('', [Validators.required, Validators.email]),
+      password: new FormControl('', [
+        Validators.required,
+        Validators.pattern('(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9]).{6,}'),
```
Comment (joacomiguens) [inline angular/src/app/screens/sign-up/sign-up.component.ts]: same

File: angular/src/app/screens/sign-up/sign-up.module.ts [hunk 25]
```diff
@@ -3,11 +3,16 @@ import { CommonModule } from '@angular/common';
 import { SignUpRoutingModule } from './sign-up-routing.module';
 import { SignUpComponent } from './sign-up.component';
 import { ReactiveFormsModule } from '@angular/forms';
```
Comment (joacomiguens) [inline angular/src/app/screens/sign-up/sign-up.module.ts]: must be at top

File: angular/src/app/screens/sign-up/sign-up.component.ts [hunk 26]
```diff
@@ -1,25 +1,41 @@
...
   templateUrl: './sign-up.component.html',
-  styleUrls: ['./sign-up.component.scss'],
 })
 export class SignUpComponent {
   fields = SIGN_UP_FIELDS;
   buttonSend = 'Sign Up';
 
   constructor(private userService: UserService) {}
 
+  formControls = new FormGroup(
```
Comment (miguelhenaoa) [inline angular/src/app/screens/sign-up/sign-up.component.ts]: You don't have a `FormControl` array, you have a variable `FormGroup` type, you can call it `formGroup` or just `form`.

File: angular/src/app/screens/sign-up/sign-up.module.ts [hunk 27]
```diff
@@ -3,11 +3,16 @@ import { CommonModule } from '@angular/common';
...
-import { FormContainerComponent } from 'src/app/components/form-container/form-container.component';
+import { FormContainerModule } from 'src/app/components/form-container/form-container.module';
```
Comment (miguelhenaoa) [inline angular/src/app/screens/sign-up/sign-up.module.ts]: same

File: angular/src/app/components/form-container/form-container.component.ts [hunk 28]
```diff
@@ -12,39 +11,30 @@ export class FormContainerComponent {
...
-      firstName: new FormControl('', [Validators.required]),
-      lastName: new FormControl('', [Validators.required]),
-      email: new FormControl('', [Validators.required, Validators.email]),
-      password: new FormControl('', [Validators.required]),
-      passwordConfirmation: new FormControl('', [Validators.required]),
-    },
-    [CustomValidators.MatchValidator('password', 'passwordConfirmation')],
-  );
+  @Input() errorMessage!: string;
+  @Input() formControls!: FormGroup;
```
Comment (miguelhenaoa) [inline angular/src/app/components/form-container/form-container.component.ts]: same

File: angular/src/app/screens/login/login.component.ts [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+  templateUrl: './login.component.html',
+})
+export class LoginComponent {
+  fields = LOGIN_FIELDS;
+  buttonSend = 'Login';
+  errorMessage!: string;
+
+  constructor(private userService: UserService) {}
+
+  formControls = new FormGroup({
```
Comment (miguelhenaoa) [inline angular/src/app/screens/login/login.component.ts]: same

**Review by miguelhenaoa — CHANGES_REQUESTED** [review]

**Review by joacomiguens — APPROVED** [review]

**Review by miguelhenaoa — APPROVED** [review]

**Pull request merged**